### PR TITLE
[SYCL][COMPAT][Doc] Add SYCLcompat documentation

### PR DESCRIPTION
**Pull request by Alcpz**
This pull request introduces a new stand alone library, `SYCLcompat`: a simplified wrapper on top of SYCL, aiming to make it more accessible to developers familiar with other heterogeneous programming models.

SYCLcompat has two primary goals:

-  Improve the adoption of SYCL. This library is designed to provide a familiar programming interface that resembles other popular heterogeneous programming models. By reducing the learning curve, it enables developers to leverage SYCL's power and features more easily.

 - Source-to-Source Translation Support. SYCLcompat is also designed to facilitate automatic source-to-source translation from other heterogeneous programming models to SYCL and offer a more standardized and consistent programming interface. This feature can significantly streamline the migration and integration of existing codebases into the SYCL ecosystem.

The first commit of this PR includes the proposed library README, providing explanation of its motivation, public interface, usage guidelines, and code examples.

A set of PRs will follow, including subsets of the current implementation including their tests.

We are open to any suggestions, concerns, or improvements you may have, so please, let us know if you have any.

Edit: Updated from extension to stand alone library.
https://github.com/intel/llvm/pull/9976 `dims.hpp` and `defs.hpp` headers

### CODE REVIEW

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_compat.asciidoc
Comment (Alcpz) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_compat.asciidoc]: > Does each of these overloads create a new SYCL kernel if multiple are called with the same template parameter F?

Our current implementation proposal compiles the kernel once, independent of which `launch<F>` overload is called by the user.
The specific code for this interface will be uploaded in a different PR. I can link that PR and mention you if you are interested in the specifics.

> Are kernels launched using these overloads always given an implementation-defined name?

Yes. Kernels are anonymous and SYCLcompat requires `-fsycl-unnamed-lambda`/ `-sycl-std=2020`, so the kernel name should be the implementation-defined name (e.g. `auto_name`)

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_compat.asciidoc
Comment (Alcpz) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_compat.asciidoc]: > Why does the kernel's function signature need to be modified? The example void local_mem_2d(int *d_A) function doesn't include a parameter for local memory.


There are two distinct interfaces to allocate local memory  using `SYCLcompat`. The one you are mentioning `local_mem_2d` uses the current implementation of _sycl_ext_oneapi_local_memory_ to provide statically sized memory allocation. In this case, the kernel function signature is not modified because it does not need an additional argument to access the local memory, the memory is defined inside the function.

The second interface to allocate local memory is used to provide dynamically sized memory through the `launch<F>` function. The user can define the size at runtime, and internally, `SYCLcompat` will inject a pointer to the local memory location as an additional argument of the kernel function. That is the reason to modify the kernel function signature.

We have modified the proposal to improve readability, since other users may arrive to the same conclusion. Sorry about the misunderstanding.

File: sycl/doc/syclcompat/README.md
Comment (zhiweij1) [inline sycl/doc/syclcompat/README.md:66]: If SYCLomatic uses this to migrate
```
dim3 a(5, 6, 7);
```
There 2 migration options:
```
// Option1
syclcompat::dim3 a(sycl::range<3>(5, 6, 7));
// Option2
syclcompat::dim3 a(7, 6, 5);
```
I think the code of option1 is redundant. And option2 is very similar to current migration:
```
sycl::range<3> a(7, 6, 5);
```
Then the only advantage we use `syclcompat::dim3` in the migration is that we can access data member directly instead of calling method .x()/.y()/.z().

Will simplifing it to
```
struct dim3 {
size_t x, y, z;
};
```
be better?

File: sycl/doc/syclcompat/README.md
Comment (zhiweij1) [inline sycl/doc/syclcompat/README.md:514]: Shall we add another memory_region enum `constant`, and set `mode = sycl::access_mode::read` if `Memory` is `constant` ?

Comment (joeatodd) [reply sycl/doc/syclcompat/README.md:514]: We modified this from the original DPCT helper header code to reflect the fact that SYCL's `constant_space` is deprecated in SYCL 2020. It depends how closely `memory_region` is aligned with SYCL's `access::address_space`. At present we have a `usm_shared` option for the `memory_region` enum, so it doesn't *exactly* reflect the `address_space`.

I can see two scenarios in which we would need/want to have a `memory_region::constant` option:
 1. We intentionally continue to support the deprecated `address_space`
 2. We have some other notion of `memory_region::constant` like "it's actually `global` but we want to use it read-only".

Do you feel that one of these situations applies here?

File: sycl/doc/syclcompat/README.md
Comment (joeatodd) [reply sycl/doc/syclcompat/README.md:66]: The key thing about `dim3` is that it is index-flipped with respect to a sycl::range. So, actually the migration of:

```
dim3 a(5, 6, 7);
```
would be:
```
syclcompat::dim3 a(5, 6, 7);
```

We agree that translating this to:
```
syclcompat::dim3 a(sycl::range<3>(7, 6, 5));
```
would be redundant, but we see two reasons to provide constructors taking a `sycl::range` (and conversion operators back to `sycl::range`):

1. Suppose a user has translated their code to `syclcompat` (either using SYCLomatic or manually), and then they want to *incrementally* migrate to standard SYCL, or interface their code with standard SYCL code. Suppose the user wants to make use of a library function which computes a `sycl::range` for their problem size, but their code uses `dim3`:

```c++
sycl::range<3> my_range = compute_3d_sycl_range(/*args*/);

syclcompat::dim3 my_dim3{my_range}; // with converting ctor
syclcompat::dim3 my_dim3{my_range[2], my_range[1], my_range[0]}; // without it

// user's syclcompat code...
```
The provided ctor is more concise & less error prone.

2. Including the definition of these ctors and conversion operators in [header code](https://github.com/intel/llvm/blob/8f00fb0ee96c9798014825043d95b1d8d981f901/sycl/include/syclcompat/dims.hpp#L35-L58) provides additional pseudo-documentation to make users aware that `dim3` and `sycl::range` are index-flipped.

In summary, we prefer to keep the additional functionality since it simplifies incremental porting to SYCL, minimizes risk of errors, and documents the `dim3` class.

File: sycl/doc/syclcompat/README.md
Comment (zhiweij1) [reply sycl/doc/syclcompat/README.md:514]: > We have some other notion of memory_region::constant like "it's actually global but we want to use it read-only".
For SYCLomatic, it can know the original memory is qualified with `__constant__`, so we can set `sycl::access_mode::read` for it to hint compiler to do more optimizations. That is the reason it want to keep `memory_region::constant`

File: sycl/doc/syclcompat/README.md
Comment (zhiweij1) [reply sycl/doc/syclcompat/README.md:66]: OK.

Then I think only one thing need note:
> So, actually the migration of:
dim3 a(5, 6, 7);
would be:
syclcompat::dim3 a(5, 6, 7);

In SYCLomatic, we have an assumption: cuda dim3 variables are only used in kernel config arguments.
So we choose to reverse the dimensions of cuda dim3 at where it is initialized/assigned, but not before submit kernel.
And according to https://github.com/intel/llvm/pull/9976/files#diff-d4bc9db1571b8e1285f6284bf3e6acd998684428de472754763c773a197c69a2R41, this constructor does not filp the index, so the migrated code by SYCLomatic will be:
```
syclcompat::dim3 a(7, 6, 5);
```

Comment (Alcpz) [reply sycl/doc/syclcompat/README.md:66]: Yes, we don't flip the indexes there. Currently in _SYCLcompat_, the `launch<function>` mechanism flips the indexes internally if you call it using a `dim3` object. So the migration of 
```
dim3 a(5, 6, 7);
```
would still be:
```
syclcompat::dim3 a(5, 6, 7);
```
and you would not need to flip the indexes if _SYCLomatic_ migrates kernel launches using _SYCLcompat_ functionalities.
As an example, SYCLomatic could translate:
```
kernel<<<grid, threads>>>(arg1, arg2, arg3);
```
as
```
syclcompat::launch<kernel>(grid, treads, arg1, arg2, arg3);
```

File: sycl/doc/syclcompat/README.md
Comment (Alcpz) [reply sycl/doc/syclcompat/README.md:514]: Yes that seems reasonable, we can add it back in.

File: sycl/doc/syclcompat/README.md
Comment (zhiweij1) [reply sycl/doc/syclcompat/README.md:66]: ok

File: sycl/doc/syclcompat/README.md
Comment (Alcpz) [reply sycl/doc/syclcompat/README.md]: @zhiweij1, We've added the constant back. Let us know if there is something else before we proceed with the merge. Thanks!

Comment (zhiweij1) [reply sycl/doc/syclcompat/README.md]: I'm OK with this PR. @AndyCHHuang @tomflinda @ziranzha @zhimingwang36 , do you have other comments?

Comment (Alcpz) [reply sycl/doc/syclcompat/README.md]: Nice. Since it's been quite some time, I'm gonna resolve this, but feel free to reach us through other means (issues,  mail) if we need to address something else. Thanks

**Comment by joeatodd** [comment]
LGTM, nice work @Alcpz

**Comment by Alcpz** [comment]
@intel/dpcpp-doc-reviewers We think this is ready for merge now.

**Comment by dm-vodopyanov** [comment]
> @intel/dpcpp-doc-reviewers We think this is ready for merge now.

@Alcpz this can't be merged now, the process is here: https://github.com/intel/llvm/blob/sycl/CONTRIBUTING.md#merge

Currently the approve from @intel/dpcpp-doc-reviewers is required.

**Comment by Alcpz** [comment]
@dm-vodopyanov Sorry for the misunderstanding. I meant the following: we consider the version good to merge,  we would like for a review from the dpcpp-doc-reviewers group

File: sycl/doc/syclcompat/README.md
Comment (bader) [inline sycl/doc/syclcompat/README.md:437]:
```suggestion
// Queries pointer allocation type
```
Is the Doxygen style comment intentional here?

File: sycl/doc/syclcompat/README.md
Comment (joeatodd) [reply sycl/doc/syclcompat/README.md:3]: We are pushing the implementation in a series of incremental PRs. The first of these is [here](https://github.com/intel/llvm/pull/9976) which simply adds a couple of small headers & their tests, and the relevant CMake.

For the structure, refer to [this](https://github.com/intel/llvm/blob/0c51f6082760f15d268bb4320c832cebe91704e9/.github/CODEOWNERS#L125) section of the CODEOWNERS.

This is not part of the DPC++ runtime library, but rather a stand-alone library which is shipped with DPC++. This structure was discussed & decided by the oneAPI Architecture Forum.

Once the first implementation PR is merged, we'll begin to put up more incremental PRs adding functionality. This approach has been taken in response to feedback from Intel requesting that we avoid monolithic PRs which are hard to review.

File: sycl/doc/syclcompat/README.md
Comment (joeatodd) [reply sycl/doc/syclcompat/README.md]: You are right, yes, and thanks for suggesting the clarification :+1:

File: sycl/doc/syclcompat/README.md
Comment (joeatodd) [reply sycl/doc/syclcompat/README.md:158]: We're a little confused by this suggestion. These examples (`local_mem_2d` here and `slope_intercept` in the Sample Code section) are not llvm/sycl headers. They are samples of code that users might write to make use of the `syclcompat` library. They don't appear anywhere in the library.

However, we do make extensive use of `template <typename T,...>` in the library headers. Is this an issue?

File: sycl/doc/syclcompat/README.md
Comment (joeatodd) [reply sycl/doc/syclcompat/README.md]: Good catch. We'll remove these elsewhere too.

**Review by joeatodd — COMMENTED** [review]
@bader thanks for your in-depth review & useful feedback. We will address the remaining issues, rebase and request another review :+1:

**Comment by joeatodd** [comment]
Hey @bader we've addressed most of your comments. I've asked for a clarification about the variable naming. Thanks!

File: sycl/doc/syclcompat/README.md
Comment (bader) [reply sycl/doc/syclcompat/README.md:168]: Okay, let's leave this issue for users to handle. I'd like to point out that this is the real issue in our code base (more details in https://github.com/intel/llvm/issues/2665, https://github.com/intel/llvm/issues/2981, https://github.com/intel/llvm/issues/5479 and https://github.com/intel/llvm/issues/6879).

**Review by bader — APPROVED** [review]

**Comment by Alcpz** [comment]
Hello @dm-vodopyanov , we have now the required approvals.

**Comment by dm-vodopyanov** [comment]
@Alcpz thanks, according to the process mentioned above, please ping faceless account next time, not the specific people.

**Pull request merged**
